recipes/flowtorch/meta.yaml

```diff
@@ -0,0 +1,57 @@
+{% set name = "flowtorch" %}
+{% set version = "0.4" %}
+
+
+package:
+  name: {{ name|lower }}
+  version: {{ version }}
+
+source:
+  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
+  sha256: 775e69535dc54615a6c92d835e29fadc8d2bcc79ba13ce3eca06ee13ec9d02a5
+
+build:
+  number: 0
+  noarch: python
+  script: {{ PYTHON }} -m pip install . -vv
+
+requirements:
+  {% set pyver = ">=3.7" %}
+  {% set torchver = ">=1.8.1" %}
+  host:
+    - python {{ pyver }}
+    # - python
+    - pip    
+    - setuptools-scm
+    # - pytorch {{ torchver }}
+    # - pytorch
+  run:
+    - python {{ pyver }}
+    - pytorch {{ torchver }}
```

Comment:
Try `pytorch-cpu` for windows here, this may solve your issue.
https://anaconda.org/conda-forge/pytorch-cpu

Reply:
```suggestion
    - pytorch-cpu {{ torchver }}
```

Reply:
(I would revert back to `-pytorch {{ torchver }}` later)

Reply:
Also as @BastianZim told you in #16963, you can ignore the failures on windows/osx when the package is noarch.

Reply:
@ngam Thank you, for the review. I recall @BastianZim mentioning about ignoring the failures on windows/osx when the package is noarch. But was trying to add the package from a different channel (`pytorch` channel), which is maintained by original repository and serves `pytorch` for all three OS's (with same version). In fact there is a way to do it as well: *at least the python code in `build_all.py` says so.*

I tried hard-coding (adding) `pytorch` to `channel_urls` in `build_all.py`, and it seems it passed the windows CI (although I am a bit skeptical about it: _the CI kept on going for a really long time and the CI run results aren't accessible_).

Reply:
You shouldn't be using any channel besides conda-forge. There is currently an issue where the only windows build available is the pytorch-cpu one.

Reply:
Looking deeper into the code base of `conda_build`, I am quite certain it has all it requires about adding more channels to install packages from.

## A few questions(QS)/observations(OB)/proposal(PP)

### A: `[QS]`

How to use this? [maintainer-docs: `conda-forge.yml:channels`](https://conda-forge.org/docs/maintainer/conda_forge_yml.html#channels)

```yaml
channels:
  # sources selects the channels to pull packages from, in order.
  sources:
    - conda-forge
    - defaults
  # targets is a list of 2-lists, where the first element is the
  # channel to push to and the second element is the label on that channel
  targets:
    - ["conda-forge", "main"]
```

### B: `[OB]` Only `build_with_mambabuild: true` is used from recipe-specific `conda-forge.yml`

I see that `build_all.py` only uses `conda-forge.yml` [here](https://github.com/conda-forge/staged-recipes/blob/a2bcbaa66db3597977da43ee92fe6b0504b9e703/.ci_support/build_all.py#L174) in `read_mambabuild()` and after reading in a recipe-specific `conda-forge.yml` file, it only evaluates `build_with_mambabuild` and finally returns `use_it: bool`. So, even if one specifies something in `recipes/<package>/conda-forge.yml`, it is currently not being read effectively. 

### C: `[QS]` _Is this advisable/possible?_

In `build_all.py:build_folders()`, while building each recipe [see here](https://github.com/conda-forge/staged-recipes/blob/a2bcbaa66db3597977da43ee92fe6b0504b9e703/.ci_support/build_all.py#L157), if the extra channels are read in from a yaml file and the list, `channel_urls` is updated, then that would add all user-specified channels to the build.

The question is, is there a mechanism of updating this to the install instructions? (Unless this is taken care of properly, a lot of confusion could prop-up down the line for maintainability and communication-to-users.)

```sh
conda install flowtorch -c conda-forge -c pytorch
```

Instead of:

```sh
conda install flowtorch -c conda-forge -c
``` 

### D: `[OB/PP]` 

The current code-base for `build_all.py` points to the use of `conda_build.api.Config` class [see here](https://github.com/conda-forge/staged-recipes/blob/a2bcbaa66db3597977da43ee92fe6b0504b9e703/.ci_support/build_all.py#L117) to fetch the config and pass it onto `conda_build.api.build` to build each recipe under `build_all.py:build_folders()` [see here](https://github.com/conda-forge/staged-recipes/blob/a2bcbaa66db3597977da43ee92fe6b0504b9e703/.ci_support/build_all.py#L157).

```python
get_config(arch, channel_urls):
    exclusive_config_file = os.path.join(conda_build.conda_interface.root_dir,
                                         'conda_build_config.yaml')
	# other code
    ...

    config = conda_build.api.Config(
        variant_config_files=variant_config_files, arch=arch,
        exclusive_config_file=exclusive_config_file, channel_urls=channel_urls,
        error_overlinking=error_overlinking)
    return config

def build_folders(recipes_dir, folders, arch, channel_urls):
    # code to contruct graphs
    ...
    
    for recipe in d.keys():
        conda_build.api.build([recipe], config=get_config(arch, channel_urls))
```

#### Proposed Change

:bulb: What I am suggesting is to make changes as follows. This would allow each of the recipes to **include _non-conda-forge_ channels** while creating the builds.

```diff
# file: build_all.py

 def build_folders(recipes_dir, folders, arch, channel_urls):
     # code to contruct graphs
     ...
     
     for recipe in d.keys():
+        # TODO: figure out how to map recipe to its folder
+		 recipe_channel_urls = get_recipe_channels(
+			 folder, 
+			 channel_urls=channel_urls, 
+			 custom_spec_file='conda-forge.yml'
+		 )
-		 conda_build.api.build([recipe], config=get_config(arch, channel_urls))
+        conda_build.api.build([recipe], config=get_config(arch, recipe_channel_urls))
```

Where the new function `get_recipe_channels()` is:

```python
def get_recipe_channels(
        folder: str, 
        channel_urls: List[str]=['local', 'conda-forge'], 
        custom_spec_file: str='conda-forge.yml',
    ) -> List[str]:
    recipe_channel_urls = []
    with open(os.path.join(folder, custom_spec_file) as f:        
        cfy = yaml.safe_load(f.read())
        for key in ['channel', 'channels']:
            # keeping similarities with: conda_build.api.get_channel_utls
            recipe_channel_urls += cfy.get(key, {}).get('sources', [])
    if recipe_channel_urls:
        # allow local paths (similar to: conda_build.api.get_channel_utls)
        recipe_channel_urls = conda_build.api.get_channel_utls(args={
            'channels': recipe_channel_urls,
        })
    else:
        recipe_channel_urls = channel_urls.copy()

    return recipe_channel_urls
```

Please let me know your thoughts. 

cc: @ngam @BastianZim @msarahan

Reply:
I think they only recently changed it so that it is only conda-forge. I am actually, like you, very new to this 😶‍🌫️  (so I am NOT a reviewer, sorry if I came across that way --- just trying to help/engage people to learn and help). So the other two people you tagged look like they know a lot more than me ;)

Reply:
@ngam I chose to go with *ignoring failing windows builds* for the time being. I have updated/removed all the files.

Reply:
> I chose to go with ignoring failing windows builds for the time being

I think that's a good idea --- you will likely hear more from experienced people about this. Usually, only the linux64 build matters for noarch packages.

Reply:
> The other two people you tagged look like they know a lot more than me ;)

What you said was completely correct :)
Conda-forge recently switched to only using the conda-forge channel: https://conda-forge.org/docs/user/announcements.html#announcements

The ability to include other channels has been kept because sometimes adding back `defaults` is required but that is only in absolute edge cases and should be discussed with conda-forge/core. It isn't necessary for this package though as it's not critical to the overall infrastructure and `noarch`, so after the review process, it doesn't matter anyways anymore.

Reply:
> pytorch-cpu

This is a relict and not really used anymore. It's best to use Pytorch and the solver will then select the best version. The [pytorch-feedstock](https://github.com/conda-forge/pytorch-cpu-feedstock) is a good place to start if you want to learn more about it but the setup is a WIP due to the difficulty of compiling PyTorch. 
In short, it's always best to just use `pytorch` currently.